packages/compass-indexes/src/modules/create-index.tsx

```diff
@@ -0,0 +1,673 @@
+      dispatch(indexCreationSucceeded(inProgressIndex.id));
+      track(
+        'Index Created',
+        trackEvent,
+        connectionInfoAccess.getCurrentConnectionInfo()
+      );
+
+      // Start a new fetch so that the newly added index's details can be
+      // loaded. indexCreationSucceeded() will remove the in-progress one, but
+      // we still need the new info.
+      await dispatch(fetchIndexes());
+    } catch (err) {
+      dispatch(indexCreationFailed(inProgressIndex.id, (err as Error).message));
+    }
+  };
+};
+
+const reducer: Reducer<State, Action> = (state = INITIAL_STATE, action) => {
+  if (isAction<FieldAddedAction>(action, ActionTypes.FieldAdded)) {
+    return {
+      ...state,
+      fields: [...state.fields, { name: '', type: '' }],
+    };
+  }
+
+  if (isAction<FieldRemovedAction>(action, ActionTypes.FieldRemoved)) {
+    const fields = [...state.fields];
+    fields.splice(action.idx, 1);
+    return {
+      ...state,
+      fields,
+    };
+  }
+
+  if (isAction<FieldTypeUpdatedAction>(action, ActionTypes.FieldTypeUpdated)) {
+    const fields = [...state.fields];
+    if (action.idx >= 0 && action.idx < fields.length) {
+      const field = { ...fields[action.idx] };
+      field.type = action.fType;
+      fields[action.idx] = field;
+    }
+    return {
+      ...state,
+      fields,
+    };
+  }
+
+  if (isAction<FieldsChangedAction>(action, ActionTypes.FieldsChanged)) {
+    return {
+      ...state,
+      fields: action.fields,
+    };
+  }
+
+  if (isAction<OptionChangedAction>(action, ActionTypes.OptionChanged)) {
+    return {
+      ...state,
+      options: {
+        ...state.options,
+        [action.name]: {
+          value: action.value,
+          // "enable" checkbox-type inputs on first change (they are not hidden
+          // behind a checkbox), otherwise keep the current value
+          enabled:
+            OPTIONS[action.name].type === 'checkbox'
+              ? true
+              : state.options[action.name].enabled,
+        },
+      },
+    };
+  }
+
+  if (isAction<OptionToggledAction>(action, ActionTypes.OptionToggled)) {
+    return {
+      ...state,
+      options: {
+        ...state.options,
+        [action.name]: {
+          ...state.options[action.name],
+          enabled: action.enabled,
+        },
+      },
+    };
+  }
+
+  if (
+    isAction<CreateIndexOpenedAction>(action, ActionTypes.CreateIndexOpened)
+  ) {
+    return {
+      ...state,
+      isVisible: true,
+    };
+  }
+
+  if (isAction<ErrorEncounteredAction>(action, ActionTypes.ErrorEncountered)) {
+    return {
+      ...state,
+      error: action.error,
+    };
+  }
+
+  if (isAction<ErrorClearedAction>(action, ActionTypes.ErrorCleared)) {
+    return {
+      ...state,
+      error: null,
+    };
+  }
+
+  if (
+    isAction<CreateIndexClosedAction>(action, ActionTypes.CreateIndexClosed)
+  ) {
+    // when hiding the modal also reset it
+    return getInitialState();
```

Comment:
Nit: usually better to do on `show` instead so that the form that is being animated out of the view by leafygreen doesn't jump weirdly due to state being reset while it's animating